### PR TITLE
Add constant folding

### PR DESCRIPTION
**Pull request by orausch**

### CODE REVIEW

File: daceml/transformation/constant_folding.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+from dace.transformation import pattern_matching
+from dace.sdfg import nodes as nd
+from dace.sdfg import utils as sdutil
+
+import daceml.onnx as donnx
+from daceml.onnx.converters import clean_onnx_name
+from daceml.onnx.nodes.onnx_op import ONNXOp
+from daceml.onnx import ONNXModel
+
+# blocklist of nondeterministic ops
```
Comment (tbennun) [inline daceml/transformation/constant_folding.py:19]: blacklist

File: daceml/transformation/constant_folding.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+                        'ONNXIf',
+                        'ONNXLoop',
+                        'ONNXMomentum',
+                        'ONNXRandomNormal',
+                        'ONNXRandomNormalLike',
+                        'ONNXRandomUniform',
+                        'ONNXRandomUniformLike',
+                        'ONNXSVMClassifier',
+                        'ONNXSVMRegressor',
+                        'ONNXScan',
```
Comment (tbennun) [inline daceml/transformation/constant_folding.py:32]: is Scan nondeterministic?

File: daceml/transformation/constant_folding.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+from dace.sdfg import utils as sdutil
+
+import daceml.onnx as donnx
+from daceml.onnx.converters import clean_onnx_name
+from daceml.onnx.nodes.onnx_op import ONNXOp
+from daceml.onnx import ONNXModel
+
+# blocklist of nondeterministic ops
+# yapf: disable
+NONDETERMINISTIC_OPS = {'ONNXGradient',
```
Comment (tbennun) [inline daceml/transformation/constant_folding.py:21]: where's dropout?

**Review by tbennun — CHANGES_REQUESTED** [review]

File: daceml/transformation/constant_folding.py (hunk 1 above)
Comment (tbennun) [reply daceml/transformation/constant_folding.py:19]: nevermind

File: setup.py [hunk 4]
```diff
@@ -23,9 +23,13 @@
...
+        'dace@git+https://github.com/spcl/dace.git@866554a', 'onnx == 1.7.0',
         'torch'
     ],
     # install with pip and --find-links (see Makefile)
     # See https://github.com/pypa/pip/issues/5898
-    extras_require={'testing': ['coverage', 'pytest', 'yapf', 'pytest-cov']})
+    extras_require={
+        'testing':
+        ['coverage', 'pytest', 'yapf', 'pytest-cov', 'transformers'],
+        'debug': ['onnxruntime']
```
[the remark this answers is not in the document]
Comment (orausch) [reply setup.py:34]: This is just to debug our ONNX graph outputs vs the ORT output (to debug if our stuff works)

File: daceml/onnx/onnx_importer.py [hunk 5]
```diff
@@ -40,13 +40,14 @@ def __init__(self,
         :param name: the name for the SDFG.
         :param model: the model to import.
         :param cuda: if `True`, weights will be passed as cuda arrays.
-        :param apply_strict: if `True`, apply strict transformations after all nodes have
+        :param apply_strict: if `True`, apply strict transformation after all nodes have
```
[the remark this answers is not in the document]
Comment (orausch) [reply daceml/onnx/onnx_importer.py:43]: I think this is a rename gone wrone

File: daceml/transformation/constant_folding.py (hunk 3 above)
Comment (orausch) [reply daceml/transformation/constant_folding.py:21]: :scream:

File: daceml/transformation/constant_folding.py (hunk 2 above)
Comment (orausch) [reply daceml/transformation/constant_folding.py:32]: it takes a graph as an argument, so I guess yes. Didn't read the whole docstring though

Comment (orausch) [reply daceml/transformation/constant_folding.py:32]: (the graph can have nondeterministic ops)

Comment (tbennun) [reply daceml/transformation/constant_folding.py:35]: ok

**Review by tbennun — APPROVED** [review]

**Pull request merged**